File: server/actualizar_capitulos.py
import os
import json
import re
from bs4 import BeautifulSoup
import requests
from urllib.parse import urljoin
import sys

uid = sys.argv[1] + '.json'


def limpiar_texto(texto):
    """Limpia el texto eliminando los símbolos no deseados"""
    texto_limpio = re.sub(r'[^\w\s]', '', texto)
    return texto_limpio.strip()

def obtener_capitulos(url):
    try:
        response = requests.get(url)
        if response.status_code == 200:
            soup = BeautifulSoup(response.content, 'html.parser')
            chapters = soup.find_all('li', class_='row')[:4]
            capitulos = []
            for chapter in chapters:
                nombre_cap = limpiar_texto(chapter.find('h4').text)
                link = urljoin(url, chapter.find('a')['href'])
                capitulos.append({'capitulo': nombre_cap, 'url': link, 'leido': False})
            return capitulos
        else:
            print(f'Error al realizar la solicitud HTTP a la URL: {url}')
            return []
    except requests.exceptions.RequestException as e:
        print(f'Error al realizar la solicitud HTTP a la URL: {url}')
        print(f'Error: {e}')
        return []

def obtener_animes(urls):
    animes = {}
    capitulos = []

    for url in urls:
        try:
            chapters = obtener_capitulos(url)
            if chapters:
                response = requests.get(url)
                if response.status_code == 200:
                    soup = BeautifulSoup(response.content, 'html.parser')
                    titulo = soup.find_all('h1')
                    imagen = soup.find_all('div', class_='media-left cover-detail')

                    base_url = response.url
                    nombre = limpiar_texto(titulo[0].text)
                    imagen_url = urljoin(base_url, imagen[0].find('img')['src'])

                    if nombre in animes:
                        # Actualizar los capítulos existentes
                        anime = animes[nombre]
                        for chapter in chapters:
                            if chapter not in anime['link']:
                                anime['link'].insert(0, chapter)
                    else:
                        animes[nombre] = {
                            'nombre': nombre,
                            'imagen_url': imagen_url,
                            'link': chapters
                        }

                    capitulos.extend([{
                        'id': len(capitulos) + 2,
                        'imagen_url': imagen_url,
                        'nombre': nombre,
                        'link': chapters
                    }])
        except requests.exceptions.RequestException as e:
            print(f'Error al realizar la solicitud HTTP a la URL: {url}')
            print(f'Error: {e}')

    return animes, capitulos



def cargar_resultados():
    if not os.path.exists(uid) or os.stat(uid).st_size == 0:
        return [], 0
    else:
        with open(uid, 'r') as archivo:
            capitulos = json.load(archivo)
        ultimo_id = max([c['id'] for c in capitulos]) if capitulos else 0
        return capitulos, ultimo_id

def actualizar_resultados(capitulos):
    with open(uid, 'w') as archivo:
        json.dump(capitulos, archivo, indent=4)
    print('Resultados guardados en resultados.json')

def main():
    urls = [
       'https://www.leercapitulo.com/manga/c4kt5e/tengoku-daimakyo/',
    ]

    animes, capitulos = obtener_animes(urls)
    resultados_capitulos, ultimo_id = cargar_resultados()

    # Actualizar los capítulos existentes con los nuevos enlaces
    for anime in resultados_capitulos:
        nombre = anime['nombre']
        if nombre in animes:
            anime['link'].extend([c for c in animes[nombre]['link'] if c not in anime['link']])

    # Agregar los nuevos capítulos a los resultados existentes
    nuevos_capitulos = [c for c in capitulos if c['id'] > ultimo_id]
    resultados_capitulos.extend(nuevos_capitulos)

    actualizar_resultados(resultados_capitulos)

if __name__ == "__main__":
    main()

File: src/app/api/actualizar/actualizar_capitulos.js
const os = require("os");
const fs = require("fs");
const axios = require("axios");
const { URL } = require("url");
const cheerio = require("cheerio");
const exp = require("constants");

const uid = process.argv[2] + ".json";

function limpiarTexto(texto) {
  // Limpia el texto eliminando los símbolos no deseados
  const textoLimpio = texto.replace(/[^\w\s]/g, "");
  return textoLimpio.trim();
}

async function obtenerCapitulos(url) {
  try {
    const response = await axios.get(url);
    if (response.status === 200) {
      const $ = cheerio.load(response.data);
      const chapters = $("li.row").slice(0, 4);
      const capitulos = [];
      chapters.each((index, element) => {
        const nombreCap = limpiarTexto($(element).find("h4").text());
        const link = new URL($(element).find("a").attr("href"), url).href;
        capitulos.push({ capitulo: nombreCap, url: link, leido: false });
      });
      return capitulos;
    } else {
      console.log(`Error al realizar la solicitud HTTP a la URL: ${url}`);
      return [];
    }
  } catch (e) {
    console.log(`Error al realizar la solicitud HTTP a la URL: ${url}`);
    console.log(`Error: ${e.message}`);
    return [];
  }
}

async function obtenerAnimes(urls) {
  const animes = {};
  const capitulos = [];

  for (const url of urls) {
    try {
      const chapters = await obtenerCapitulos(url);
      if (chapters.length > 0) {
        const response = await axios.get(url);
        if (response.status === 200) {
          const $ = cheerio.load(response.data);
          const titulo = $("h1");
          const imagen = $("div.media-left.cover-detail");

          const baseUrl = response.config.url;
          const nombre = limpiarTexto($(titulo[0]).text());
          const imagenUrl = new URL(
            $(imagen[0]).find("img").attr("src"),
            baseUrl
          ).href;

          if (animes[nombre]) {
            // Actualizar los capítulos existentes
            const anime = animes[nombre];
            for (const chapter of chapters) {
              if (!anime.link.some((c) => c.capitulo === chapter.capitulo)) {
                anime.link.unshift(chapter);
              }
            }
          } else {
            animes[nombre] = {
              nombre: nombre,
              imagenUrl: imagenUrl,
              link: chapters,
            };
          }

          capitulos.push({
            id: capitulos.length + 2,
            imagenUrl: imagenUrl,
            nombre: nombre,
            link: chapters,
          });
        }
      }
    } catch (e) {
      console.log(`Error al realizar la solicitud HTTP a la URL: ${url}`);
      console.log(`Error: ${e.message}`);
    }
  }

  return [animes, capitulos];
}

function cargarResultados() {
  if (!fs.existsSync(uid) || fs.statSync(uid).size === 0) {
    return [[], 0];
  } else {
    const capitulos = require(`./${uid}`);
    const ultimoId = capitulos.reduce((maxId, c) => Math.max(maxId, c.id), 0);
    return [capitulos, ultimoId];
  }
}

function actualizarResultados(capitulos) {
  fs.writeFileSync(uid, JSON.stringify(capitulos, null, 4));
  console.log("Resultados guardados en resultados.json");
}

async function main() {
  const urls = [
    "https://www.leercapitulo.com/manga/c4kt5e/tengoku-daimakyo/",
    "https://www.leercapitulo.com/manga/5d4b5e/one-piece/",
    "https://www.leercapitulo.com/manga/5q8ord/boku-no-kokoro-no-yabai-yatsu/",
  ];

  const [animes, capitulos] = await obtenerAnimes(urls);
  const [resultadosCapitulos, ultimoId] = cargarResultados();

  // Actualizar los capítulos existentes con los nuevos enlaces
  for (const anime of resultadosCapitulos) {
    const nombre = anime.nombre;
    if (nombre in animes) {
      anime.link.push(
        ...animes[nombre].link.filter(
          (c) => !anime.link.some((ac) => ac.capitulo === c.capitulo)
        )
      );
    }
  }

  // Agregar los nuevos capítulos a los resultados existentes
  const nuevosCapitulos = capitulos.filter((c) => c.id > ultimoId);
  resultadosCapitulos.push(...nuevosCapitulos);

  actualizarResultados(resultadosCapitulos);
}

main();
